src/signed.rs

```diff
@@ -0,0 +1,68 @@
+use crate::{Index, KeyBox};
+use codec::{Decode, Encode};
+
+pub trait Signable {
+    fn bytes_to_sign(&self) -> Vec<u8>;
+}
+
+#[derive(Clone, Debug, Decode, Encode)]
+pub struct UncheckedSigned<T: Signable + Index, S> {
+    pub signable: T,
+    pub signature: S,
+}
+
+#[derive(Debug)]
+pub struct Signed<'a, T: Signable + Index, KB: KeyBox> {
+    pub unchecked: UncheckedSigned<T, KB::Signature>,
+    pub key_box: &'a KB,
+}
+
+impl<'a, T: Signable + Index + Clone, KB: KeyBox> Clone for Signed<'a, T, KB> {
+    fn clone(&self) -> Self {
+        Signed {
+            unchecked: self.unchecked.clone(),
+            key_box: self.key_box,
+        }
+    }
+}
+
+impl<'a, T: Signable + Index, KB: KeyBox> Signed<'a, T, KB> {
+    pub fn sign(key_box: &'a KB, signable: T) -> Self {
+        let signature = key_box.sign(&signable.bytes_to_sign());
```

Comment:
We should check here that `signable.index()` agrees with our index – I'm not sure whether the `KeyBox` is `Index`, but it should be.

Reply:
It does implement `Index` buti I don't think it should...

Reply:
It only has one private key, the one corresponding to `Index`, so I think the meaning is reasonable. Here it adds the possibility for some validation – if for whatever reason we tried signing something created by someone else, we would get an explicit error instead of massive confusion in the consensus algorithm when someone sends such an object.

If you don't want that check here then maybe implement this signing procedure only for `T` without `Index`? Then the semantics would be clear again – just sign the bytes, there cannot be a mismatch with an index if the index is not officially there.

Reply:
The type `T` is no longer required to implement the `Index` trait, so we don't have to ensure that the indices match